Enable Maintenance Stats for Bitbucket Projects

mikeyoung85 opened the pull request:
Enables maintenance stats for projects with Bitbucket repositories. Gathers up stats for commits, issues, and pull requests.

katzj commented on app/models/gather_repository_maintenance_stats.rb:
```diff
@@ -1,12 +1,33 @@
 class GatherRepositoryMaintenanceStats
+    def self.gather_bitbucket_stats(repository)
```
Would some of this make more sense in the repository hosts instead of having `gather_bitbucket` and `gather_github`?

mikeyoung85 replied:
I think that would make more sense now that I've gone through `repository_host` with this work. Then the API to get stats would just be `repository.gather_stats`. I'll shift this stuff over today.

katzj commented on app/models/repository_host/bitbucket.rb:
```diff
@@ -70,6 +70,10 @@ def download_pull_requests(token = nil)
     end
 
+    def retrieve_commits(token = nil)
+      api_client(token).repos.commits.list(repository.owner_name, repository.project_name)['values']
```
How doe sthis work with rate limiting?

mikeyoung85 replied:
This is the only documentation I can find on rate limits: https://confluence.atlassian.com/bitbucket/rate-limits-668173227.html?_ga=2.192562154.1743157817.1561468931-1728484474.1527255352. Looks to be 1000 requests/hour on the repositories and commits. It looks like the rate is tied to the IP address if unauthenticated and the user if it is authenticated.

katzj commented on app/models/repository_issue/bitbucket.rb:
```diff
@@ -40,7 +40,7 @@ def self.create_from_hash(name_with_owner, issue_hash, token = nil)
       i.repository_id = repository.id
       i.labels = [issue_hash.fetch(:metadata, {})[:kind]]
-      i.pull_request = issue_hash[:type] == 'pull_request'
+      i.pull_request = issue_hash[:type] == 'pullrequest'
```
Is this a bug that's existed for a whiel?

mikeyoung85 replied:
It's something I noticed while working through this. I'm not sure when that value changed, but it was what I was seeing in the live API responses. We also have a mix of V1 and V2 API calls to Bitbucket and it could differ between the two?

kbarrette commented on app/models/gather_repository_maintenance_stats.rb:
```diff
@@ -1,12 +1,33 @@
     def self.gather_stats(repository)
         # only support Github repos for now
         # check to make sure the Project URLs are also pointing to a Github repository
         unless stats_enabled?(repository)
```
I think this might be easier to read as something like:
```ruby
if github?(repository)
  gather_github_stats(repository)
elsif bitbucket?(repository)
  gather_bitbucket_stats(repository)
else
  repository.repository_maintenance_stats.destroy_all
  nil
end
```

kbarrette commented on app/models/maintenance_stats/stats/bitbucket/commit_count_stats.rb:
```diff
@@ -0,0 +1,21 @@
+module MaintenanceStats
+    module Stats
```
Too much indentation!

kbarrette commented on app/models/maintenance_stats/stats/bitbucket/commit_count_stats.rb:
```diff
@@ -0,0 +1,21 @@
+        module Bitbucket
+            class CommitsStat < MaintenanceStats::Stats::BaseStat
+                def count_up_commits(dataset, since)
+                    return nil if dataset.nil? || dataset.empty?
```
`return unless dataset.present?`

kbarrette commented on app/models/maintenance_stats/stats/bitbucket/commit_count_stats.rb:
```diff
@@ -0,0 +1,21 @@
+
+                def get_stats
+                    {
+                        "last_week_commits": count_up_commits(@results, 1.week.ago),
```
Shouldn't need the quotes

kbarrette commented on app/models/maintenance_stats/stats/bitbucket/commit_count_stats.rb:
```diff
@@ -0,0 +1,21 @@
+                def get_stats
+                    {
+                        "last_week_commits": count_up_commits(@results, 1.week.ago),
+                        "last_month_commits": count_up_commits(@results, 1.month.ago),
```
Do we want to reference these all from the same time? It's possible to roll over a day or whatever and get unexpected results

kbarrette commented on app/models/maintenance_stats/stats/bitbucket/issue_stats.rb:
```diff
@@ -0,0 +1,81 @@
+                    return @closed_issues_count unless @closed_issues_count.nil?
+            
+                    @closed_issues_count = count_issues(issues, state: "closed")
+                    @closed_issues_count
```
No need for this line - the value of the assignment is the assigned value

The usual pattern would be something like:
```ruby
@closed_issues_count ||= count_issues(issues, state: "closed")

kbarrette commented on app/models/maintenance_stats/stats/bitbucket/issue_stats.rb:
```diff
@@ -0,0 +1,81 @@
+            
+                  def issue_closure_rate
+                    return 1.0 if count_issues(issues) == 0
+                    closed_issues_count.to_f / count_issues(issues).to_f
```
Should we be using `BigDecimal` here for more precision, or is floating point OK?

mikeyoung85 replied:
I think floating should be fine, precision isn't a big factor here.

mikeyoung85 commented:
Updated the new code with the suggested changes. I also moved the stats gathering under the correct `repository_host` class and now gathering is called from the repository directly, which calls the correct set of queries and stats. I also moved the Github stats under their own module to keep the two sets distinct from each other. I would like to also move the stats that way, but that will require some effort to redo the VCR mocks since those have class names embedded in the calls.

kbarrette commented on app/models/repository_host/github.rb:
```diff
@@ -221,10 +225,84 @@ def download_tag(token, tag, existing_tag_names)
+      response.errors.each do |message|
+          Rails.logger.warn(message)
+      end
+      response.data.errors.each do |message|
```
You can one-line this like
```ruby
response.data.errors.each(&Rails.logger.method(:warn)) if response.data.errors.present?
```

kbarrette commented on app/models/maintenance_stats/stats/github/commit_count_stats.rb:
```diff
@@ -0,0 +1,61 @@
+        end
+      end
+
+      class LastWeekCommitsStat < BaseCommitCount
```
Out of curiosity, why classes with a single method in them rather than a bunch of methods?

mikeyoung85 replied:
I think it was initially in the way I set up the method that took queries and assigned them to stat classes. That was changed before merge to be a little more simple. I would like to run back through the Github pieces tomorrow and get those queries and stat classes cleaned up.
